### PR TITLE
Martin/new message detail p2

### PR DESCRIPTION
**Pull request by MartinJibber**
This fixes multiple issues causing the message consumption to not happen, or happen at the wrong time.

### CODE REVIEW

File: Shared/Models/Message Model/Messageable.swift [hunk 1]
```diff
@@ -56,7 +56,7 @@ func ==(lhs: Messageable, rhs: Messageable) -> Bool {
 extension Messageable {
 
     var canBeConsumed: Bool {
-        return !self.isConsumedByMe && self.isFromCurrentUser
+        return !self.isConsumedByMe && !self.isFromCurrentUser
```
Comment (MartinJibber) [inline Shared/Models/Message Model/Messageable.swift:59]: We only want to consume messages if they're not already consumed, and they're not from the current user.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 2]
```diff
@@ -18,7 +18,7 @@ protocol ConversationUIStateSettable {
 /// A cell to display the messages of a conversation.
 /// The user's messages and other messages are put in two stacks (along the z-axis),
 /// with the most recent messages at the front.
-class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettable {
+class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettable, UICollectionViewDelegate {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:21]: The messages cell needs to know when messages appear so it know when there's a new frontmost message.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 3]
```diff
@@ -29,7 +29,7 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettabl
     var handleTappedConversation: ((MessageSequence) -> Void)?
     var handleDeleteConversation: ((MessageSequence) -> Void)?
 
-    @Published var incomingTopmostMessage: ChatMessage?
+    @Published var frontmostNonUserMessage: ChatMessage?
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:32]: Slight naming clarification.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 4]
```diff
@@ -89,12 +89,8 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettabl
         }
 
         self.dataSource.handleLoadMoreMessages = { [unowned self] cid in
-            Task {
-                self.conversationController?.loadPreviousMessages()
-            }
+            self.conversationController?.loadPreviousMessages()
         }
-        
-        self.collectionLayout.delegate = self
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:97]: The messages layout no longer has a delegate. Unfortunately, the delegate would sometimes fire before the cell had appeared which caused problems.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 5]
```diff
@@ -218,6 +214,28 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettabl
...
+                          messageCell.areDetailsShown else { continue }
+
+                    guard let item = self.dataSource.itemIdentifier(for: indexPath),
+                          case .message(let cid, let messageID, _) = item else { continue }
+
+                    let message = ChatClient.shared.message(cid: cid, id: messageID)
+
+                    self.startConsumptionIfNeeded(for: message, at: indexPath)
+                }
+            }.store(in: &self.subscriptions)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift]: Listen for changes to the active conversation to determine if we need to start consuming messages.

This fixes issues where sliding to a new conversation didn't start consuming a message until you scrolled.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 6]
```diff
@@ -33,6 +38,14 @@ class MessageCell: UICollectionViewCell {
...
     private func initializeViews() {
         self.contentView.addSubview(self.content)
+
+        ConversationsManager.shared.$activeConversation
+            .removeDuplicates()
+            .mainSink { [unowned self] activeConversation in
+                // If this cell's conversation becomes active,
+                // then start message consumption if needed.
+                self.handleDetailsShown(self.detailVC.view.alpha == 1)
+            }.store(in: &self.subscriptions)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift:48]: This fixes the issue where a message wouldn't start being consumed when you slid over to a new conversation.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 7]
```diff
@@ -102,35 +111,43 @@ class MessageCell: UICollectionViewCell {
...
 
-    func handleConsumption() {
-        guard ChatUser.currentUserRole != .anonymous,
-              let message = self.messageState.message,
-              message.canBeConsumed else {
-                  return
-              }
+    // MARK: - Message Consumption
 
+    private var consumeMessageTask: Task<Void, Never>?
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift:121]: Once we start consumption, we may need to cancel it.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 8]
```diff
@@ -102,35 +111,43 @@ class MessageCell: UICollectionViewCell {
...
+
+    private func handleDetailsShown(_ areDetailsShown: Bool) {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift:123]: Message consumption logic.

**Review by Benjamindodgson — APPROVED** [review]
Awesome work!

**Pull request merged**